cmd/dryrun/dryrun.go

```diff
@@ -0,0 +1,479 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	ocp_configv1 "github.com/openshift/api/config/v1"
+	"github.com/portworx/sched-ops/k8s/apiextensions"
+	coreops "github.com/portworx/sched-ops/k8s/core"
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	log "github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	fakeextclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/fake"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	fakek8sclient "k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/record"
+	cluster_v1alpha1 "sigs.k8s.io/cluster-api/pkg/apis/deprecated/v1alpha1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/yaml"
+
+	"github.com/libopenstorage/operator/drivers/storage"
+	_ "github.com/libopenstorage/operator/drivers/storage/portworx"
+	"github.com/libopenstorage/operator/drivers/storage/portworx/component"
+	pxutil "github.com/libopenstorage/operator/drivers/storage/portworx/util"
+	"github.com/libopenstorage/operator/pkg/apis"
+	corev1 "github.com/libopenstorage/operator/pkg/apis/core/v1"
+	"github.com/libopenstorage/operator/pkg/controller/storagecluster"
+	_ "github.com/libopenstorage/operator/pkg/log"
+	testutil "github.com/libopenstorage/operator/pkg/util/test"
+	"github.com/libopenstorage/operator/pkg/version"
+	inttestutil "github.com/libopenstorage/operator/test/integration_test/utils"
+)
+
+const (
+	flagVerbose       = "verbose"
+	flagStorageCluser = "storagecluster,stc"
+	flagKubeConfig    = "kubeconfig"
+	flagOutputFile    = "output,o"
+)
+
+var (
+	defaultOutputFile = "portworxCompnentSpecs.yaml"
+	skipComponents    = map[string]bool{
+		// CRD component registration is stuck at CRD validation as fake client does not set status of CRD.
+		// Potential fixes:
+		//   1. remove the validation -- need more tests to see if it's safe to remove it.
+		//   2. create a thread to monitor the CRD and update status for dry run.
+		component.PortworxCRDComponentName: true,
+		// Disruption budget component needs to talk to portworx SDK, it also needs real k8s cluster to set correct value.
+		component.DisruptionBudgetComponentName: true,
+		// Alert manager requires a secret to be deployed first.
+		// We may read the secret from real k8s cluster or generate a secret for dry run.
+		component.AlertManagerComponentName: true,
```

Comment:
We should create a fake secret in the fake client and then skip that when adding it to the output file.

Reply:
i will fix in the end, after the migration spec validation pr.